cmd/serverless-init/main.go

```diff
@@ -8,108 +8,126 @@
-	"github.com/DataDog/datadog-agent/cmd/serverless-init/initcontainer"
-	"github.com/DataDog/datadog-agent/cmd/serverless-init/log"
 	"github.com/DataDog/datadog-agent/cmd/serverless-init/metric"
-	"github.com/DataDog/datadog-agent/cmd/serverless-init/tag"
+	serverlessInitTag "github.com/DataDog/datadog-agent/cmd/serverless-init/tag"
 	logsAgent "github.com/DataDog/datadog-agent/comp/logs/agent"
-	"github.com/DataDog/datadog-agent/pkg/config"
+	pkgconfig "github.com/DataDog/datadog-agent/pkg/config"
 	"github.com/DataDog/datadog-agent/pkg/config/model"
 	configUtils "github.com/DataDog/datadog-agent/pkg/config/utils"
 	"github.com/DataDog/datadog-agent/pkg/serverless/metrics"
 	"github.com/DataDog/datadog-agent/pkg/serverless/otlp"
 	"github.com/DataDog/datadog-agent/pkg/serverless/random"
-	"github.com/DataDog/datadog-agent/pkg/serverless/tags"
+	serverlessTag "github.com/DataDog/datadog-agent/pkg/serverless/tags"
 	"github.com/DataDog/datadog-agent/pkg/serverless/trace"
 	tracelog "github.com/DataDog/datadog-agent/pkg/trace/log"
 	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
-	logger "github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-const (
-	datadogConfigPath = "/var/task/datadog.yaml"
-	logLevelEnvVar    = "DD_LOG_LEVEL"
-	loggerName        = "SERVERLESS_INIT"
-)
+const datadogConfigPath = "datadog.yaml"
 
-type cliParams struct {
-	args []string
-}
+var modeConf mode.Conf
 
 func main() {
-	if len(os.Args) < 2 {
-		panic("[datadog init process] invalid argument count, did you forget to set CMD ?")
-	}
 
-	cliParams := &cliParams{
-		args: os.Args[1:],
-	}
-
-	err := fxutil.OneShot(run, fx.Supply(cliParams))
+	modeConf = mode.DetectMode()
+	setEnvWithoutOverride(modeConf.EnvDefaults)
+
+	err := fxutil.OneShot(
+		run,
+		autodiscoveryimpl.Module(),
+		workloadmetafx.Module(),
+		fx.Provide(func(config coreconfig.Component) healthprobeDef.Options {
+			return healthprobeDef.Options{
+				Port:           config.GetInt("health_port"),
+				LogsGoroutines: config.GetBool("log_all_goroutines_when_unhealthy"),
+			}
+		}),
+		taggerimpl.Module(),
+		healthprobeFx.Module(),
+		fx.Supply(workloadmeta.NewParams()),
+		fx.Supply(tagger.NewTaggerParams()),
+		fx.Supply(core.BundleParams{
+			ConfigParams: coreconfig.NewParams("", coreconfig.WithConfigMissingOK(true)),
+			SecretParams: secrets.NewEnabledParams(),
+			LogParams:    logimpl.ForOneShot(modeConf.LoggerName, "off", true)}),
+		core.Bundle(),
+	)
 
 	if err != nil {
-		logger.Error(err)
+		log.Error(err)
+		os.Exit(-1)
 	}
 }
 
-func run(cliParams *cliParams) {
-	cloudService, logConfig, traceAgent, metricAgent, logsAgent := setup()
-	initcontainer.Run(cloudService, logConfig, metricAgent, traceAgent, logsAgent, cliParams.args)
-}
+// removing these unused dependencies will cause silent crash due to fx framework
+func run(_ secrets.Component, _ autodiscovery.Component, _ healthprobeDef.Component) {
```

Comment:
We could provide the config component to the run function and later to the setup function, and do not relay on confog.Datadog(), but the actual config component.